### PR TITLE
Fix renaming crates as they come from 2 sources

### LINKED ISSUE #5413: Renaming dependencies fails with same crate.
The following configuration:

```toml
cargo-features = ["rename-dependency"]

[package]
name = "test"
version = "0.1.0"
authors = []

[dependencies]
foo = "0.1"
bar = { git = "https://github.com/foo/bar", package = "foo" }
```

fails with:

```
Dependency 'foo' has different source paths depending on the build target. Each dependency must have a single canonical source path irrespective of build target.
```

EDIT: minimized example.

### PR DESCRIPTION
**Pull request by alexcrichton**
Previously there was a verification in manifest parsing that the same dependency
must come from the same source, but this erroneously triggered an error to get
emitted when the `package` key was used to rename crates. The first change here
was to update that clause to key off the `rename` field rather than the `name`
field.

Afterwards, though, this exposed an existing bug in the implementation. During
compilation we have a `Resolve` which is a graph of crates, but we don't know
*why* each edge in the dependency graph exists. In other words we don't know,
when looking at an edge of the graph, what `Dependency` caused that edge to be
drawn. We need to know this when passing `--extern` flags because the
`Dependency` is what lists what's being renamed.

This commit then primarily refactors `Resolve::deps` from an iterator of package
ids to an iterator of a tuples. The first element is the package id from before
and the second element is a list of `Dependency` directives which caused the
edge to ber driven.

This refactoring cleaned up a few places in the backend where we had to work
around the lack of this knowledge. Namely this also fixes the extra test added
here.

Closes #5413

### CODE REVIEW

File: src/cargo/core/resolver/resolve.rs [hunk 1]
```diff
@@ -19,21 +18,47 @@ use super::encode::Metadata;
...
 pub struct Resolve {
-    pub graph: Graph<PackageId>,
-    pub replacements: HashMap<PackageId, PackageId>,
-    pub empty_features: HashSet<String>,
-    pub features: HashMap<PackageId, HashSet<String>>,
-    pub checksums: HashMap<PackageId, Option<String>>,
-    pub metadata: Metadata,
-    pub unused_patches: Vec<PackageId>,
+    graph: Graph<PackageId>,
+    dependencies: HashMap<(PackageId, PackageId), Vec<Dependency>>,
```
Comment (matklad) [inline src/cargo/core/resolver/resolve.rs]: Hm, naively I would think that edge data would be stored as, well, edges in the graph. 

Perhaps we should refactor `pub struct Graph<N>` to

```
pub struct Graph<V, E> {
   edges: HashMap<V, HashMap<V, E>>
}
```

and instantiate it as `Graph<PackageId, Vec<Dependency>>`.

If we go down that road, then perhaps let's explore using `petgraph` for it? rust-lang/rust already depends on it, so it won't be a too new dependency.

File: src/cargo/ops/cargo_compile.rs [hunk 2]
```diff
@@ -397,8 +397,7 @@ pub fn compile_ws<'a>(
 
         // Include features enabled for use by dependencies so targets can also use them with the
         // required-features field when deciding whether to be built or skipped.
-        let deps = resolve_with_overrides.deps(package_id);
-        for dep in deps {
+        for (dep, _) in resolve_with_overrides.deps(package_id) {
             for feature in resolve_with_overrides.features(dep) {
                 features.insert(dep.name().to_string() + "/" + feature);
```
Comment (matklad) [inline src/cargo/ops/cargo_compile.rs]: Should we use renamed name here as well? So that, for example, you can specify different features for different major versions of the same package?

File: src/cargo/ops/cargo_output_metadata.rs [hunk 3]
```diff
@@ -109,7 +109,7 @@ where
         .iter()
         .map(|id| Node {
             id,
-            dependencies: resolve.deps(id).collect(),
+            dependencies: resolve.deps(id).map(|p| p.0).collect(),
```
Comment (matklad) [inline src/cargo/ops/cargo_output_metadata.rs]: Hm, we really need to surface information about renames into the metadata. A backwards compatible way to do that would be to add a new array, `dependency_info`, alongside `dependencies`, with the same length.

File: src/cargo/util/toml/mod.rs [hunk 4]
```diff
@@ -743,7 +743,8 @@ impl TomlManifest {
             let mut names_sources = BTreeMap::new();
             for dep in &deps {
                 let name = dep.name();
-                let prev = names_sources.insert(name, dep.source_id());
+                let name = dep.rename().unwrap_or(&name);
```
Comment (matklad) [inline src/cargo/util/toml/mod.rs]: Let's inline the first `let name` here?

```
let name = dep.rename().unwrap_or(dep.name().as_str());
```

File: tests/testsuite/rename_deps.rs [hunk 5]
```diff
@@ -145,3 +147,135 @@ fn rename_with_different_names() {
...
+            "#,
+        )
+        .file("foo/src/lib.rs", "pub fn foo() {}")
+        .build();
+
+    assert_that(
+        p.cargo("build -v").masquerade_as_nightly_cargo(),
+        execs().with_status(0),
+    );
+}
```
Comment (matklad) [inline tests/testsuite/rename_deps.rs]: In the implementation, we use `Vec<Dependency>` to store dependencies between two packages. Why do we need a `Vec` and not just a `Depedendency`? Do we allow to depend on the same crate under different names? Or is it because a dependency might be duplicated in `[dependencies]` and `[build-dependnecies]`? Is `rustc` ok with several `--extern` pointing to the same crate? Let's have a test for it? 

Ah, now I see it is handled in code by `multiple dependencies listed for the same crate` error, so let's have a test for exactly that error message!

Comment (matklad) [reply tests/testsuite/rename_deps.rs]: Speaking of tests, let's also add a test that checks that if we rename dependency in Cargo.toml, then we'll rebuild the code? Looking at the fingerprint calculation, looks like we might not do it? Perhaps this is a separate issue unrelated to the PR though :)

File: src/cargo/core/compiler/context/unit_dependencies.rs [hunk 6]
```diff
@@ -64,54 +64,55 @@ fn compute_deps<'a, 'b, 'cfg>(
 
     let id = unit.pkg.package_id();
     let deps = cx.resolve.deps(id);
-    let mut ret = deps.filter(|dep| {
-        unit.pkg
-            .dependencies()
-            .iter()
-            .filter(|d| d.name() == dep.name() && d.version_req().matches(dep.version()))
-            .any(|d| {
```
Comment (matklad) [inline src/cargo/core/compiler/context/unit_dependencies.rs]: Nice simplification here!

File: src/cargo/core/compiler/mod.rs [hunk 7]
```diff
@@ -1094,23 +1094,35 @@ fn build_deps_args<'a, 'cfg>(
...
+                    (None, None) => continue,
+                    (Some(n), Some(prev)) if n == prev => continue,
+                    (None, Some(n)) if n == crate_name => continue,
+                    (Some(n), None) => {
+                        name = Some(n);
+                        continue
+                    }
+                    _ => {}
+                }
+                bail!("multiple dependencies listed for the same crate must \
```
Comment (matklad) [inline src/cargo/core/compiler/mod.rs]: I believe this is correct, but the logic here is kinda hard to follow. Perhaps the following would be easier to undestand?

```
let mut names = deps.iter().map(|d| d.rename().unwrap_or(crate_nam));
let name = names.next().unwrap();
for n in names {
    if n != name { bail!("multiple dependencies...") }
}
```

File: src/cargo/core/lockfile.rs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+        Ok(Resolve {
+            graph: g,
+            empty_features: HashSet::new(),
+            features: HashMap::new(),
+            replacements,
+            checksums,
+            metadata,
+            unused_patches,
+        })
+}
```
Comment (matklad) [inline src/cargo/core/lockfile.rs]: This does not belong here probably :)

File: src/cargo/core/resolver/resolve.rs (hunk 1 above)
Comment (matklad) [reply src/cargo/core/resolver/resolve.rs]: >The problem is that lock files are then a different graph data structure (no Dependency information) than the Resolve itself (which has Dependency information). That turned out to be a much larger refactoring than necessary so I instead kept the edge data in a side table

Ok, though I don't feel that lockfiles should be a particular problem? For lockfiles, we could store the empty `Vec` of edges? I.e., [this line](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/encode.rs#L109) should be changed from `g.link(id.clone(), to_depend_on);` to

```
// Note: we don't store edges info in the lockfile, so using an empty Vec here
g.link(id.clone(), to_depend_on, Vec::new());
```

That might cause problems elsewhere of course :)

Comment (matklad) [reply src/cargo/core/resolver/resolve.rs]: Hm, let's merge this implementation, and I'll look into probably petgraph separately?

File: src/cargo/ops/cargo_output_metadata.rs (hunk 3 above)
Comment (matklad) [reply src/cargo/ops/cargo_output_metadata.rs]: I'll do that!

File: src/cargo/ops/cargo_compile.rs (hunk 2 above)
Comment (alexcrichton) [reply src/cargo/ops/cargo_compile.rs]: Hm I'm not sure, I'll log this on the tracking issue.

File: tests/testsuite/rename_deps.rs (hunk 5 above)
Comment (matklad) [reply tests/testsuite/rename_deps.rs]: >Speaking of tests, let's also add a test that checks that if we rename dependency in Cargo.toml, then we'll rebuild the code?

Added a test in https://github.com/rust-lang/cargo/pull/5425. Looks like we indeed don't handle it yet, will look into fixing that!

Comment (alexcrichton) [reply tests/testsuite/rename_deps.rs]: Er oops sorry meant to do this and forgot, thanks though for picking it up!
